src/compiler/program.ts

```diff
@@ -2266,8 +2266,13 @@ namespace ts {
 
             let redirectedPath: Path | undefined;
             if (refFile) {
-                const redirect = getProjectReferenceRedirect(fileName);
-                if (redirect) {
+                const redirectProject = getProjectReferenceRedirectProject(fileName);
+                if (redirectProject) {
+                    if (redirectProject.commandLine.options.outFile || redirectProject.commandLine.options.out) {
+                        // Shouldnt create many to 1 mapping file in --out scenario
+                        return undefined;

```

Comment:
Trying to see if I understand:

Normally, if you call `findSourceFile` for say `subproject/a.ts` which is part of a referenced project, you'd return (after updating caches) the source file for `subproject/a.d.ts`. But if that referenced project has an `outFile` option, there won't be a `subproject/a.d.ts`, there will be a single concatenated definition file, let's say `subproject/built.d.ts`.

So you return early here because `subproject/built.d.ts` has already been processed, and doing it again causes problems. That makes sense to me, but why not return the SourceFile for `subproject/built.d.ts`?

Reply:
You do not want to return `subproject/built.d.ts` because normally file is 1:1 mapping and breaking that could lead to more severe breaks down the line somewhere. Technically there is no `single` file with `subproject/a.ts` but its mix of multiple files in subproject